### PR TITLE
Adding cgroups path to the Spec.

### PR DESCRIPTION
**Pull request by vishh**
Related to #114
This change should be useful for exec'ing new processes in an existing container.

### CODE REVIEW

File: runtime-config-linux.md [hunk 1]
```diff
@@ -130,6 +130,13 @@ Next parameters can be specified:
 Also known as cgroups, they are used to restrict resource usage for a container and handle
 device access.  cgroups provide controls to restrict cpu, memory, IO, and network for
 the container. For more information, see the [kernel cgroups documentation](https://www.kernel.org/doc/Documentation/cgroups/cgroups.txt).
+The path to the cgroups can be specified in the Spec. The path is expected to be relative
+to the cgroups mount point. The Spec does not support split hierarchy. The cgroups will be 
+created if they don't exist.
```
Comment (wking) [inline runtime-config-linux.md]: This should be [one line per sentence][1].  I'll file a PR suggesting we move stuff like that from the README to a `CONTRIBUTING.md`

[1]: https://github.com/opencontainers/specs/tree/138deee1418d2376a4e9d4966e9f1543e717a641#markdown-style

File: runtime-config-linux.md [hunk 2]
```diff
@@ -130,6 +130,13 @@ Next parameters can be specified:
 Also known as cgroups, they are used to restrict resource usage for a container and handle
 device access.  cgroups provide controls to restrict cpu, memory, IO, and network for
 the container. For more information, see the [kernel cgroups documentation](https://www.kernel.org/doc/Documentation/cgroups/cgroups.txt).
+The path to the cgroups can be specified in the Spec. The path is expected to be relative
+to the cgroups mount point. The Spec does not support split hierarchy. The cgroups will be 
+created if they don't exist. 
+
+```json
+   "cgroupsPath": "/myRuntime/myContainer"
```
Comment (wking) [inline runtime-config-linux.md]: I'd mirror [cgexec(1)][1] with something like:

    "cgroups": {
      "join": [
        {
          "path": "/myRuntime/test1",
          "controllers": ["cpu", "memory"]
        },
        {
          "path": "/myRuntime/test2",
          "controllers": ["swap"]
        }
      ]
    }

And it would be nice to move the current "resources" into:

    "cgroups": {
      "create": …,
    }

[1]: http://linux.die.net/man/1/cgexec

**Comment by wking** [comment]
This might be worth floating on the list first [1] ;).  Reviewing
different proposals (e.g. your path in be6ec5ff vs. my array of
objects in [2]) is probably too much to wedge into GitHub's inline
comments :p.

[1]: https://github.com/opencontainers/specs/tree/138deee1418d2376a4e9d4966e9f1543e717a641#discuss-your-design
[2]: https://github.com/opencontainers/specs/pull/137/files#r38597501

File: runtime-config-linux.md (hunk 2 above)
Comment (vishh) [inline runtime-config-linux.md]: Won't this break with [unified hierarchy](https://www.kernel.org/doc/Documentation/cgroups/unified-hierarchy.txt)?

**Comment by vishh** [comment]
@wking: Apologies for not being updated on the new contribution workflows.

@mrunalp: Any thoughts on systemd compatibility?

File: runtime-config-linux.md (hunk 2 above)
Comment (wking) [inline runtime-config-linux.md]: On Wed, Sep 02, 2015 at 05:09:05PM -0700, Vish Kannan wrote:
> Won't this break with [unified
> hierarchy](https://www.kernel.org/doc/Documentation/cgroups/unified-hierarchy.txt)?

Ah, that's interesting reading.  In that case, a single path is fine.

Can we link to those Kernel docs from our “The Spec does not support
split hierarchy.”?

File: runtime_config_linux.go [hunk 3]
```diff
@@ -21,6 +21,9 @@ type LinuxRuntime struct {
 	// Resources contain cgroup information for handling resource constraints
 	// for the container
 	Resources Resources `json:"resources"`
+	// CgroupsPath specifies the path to cgroups that are created and/or joined by the container.
+	// The path is expected to be relative to the cgroups mountpoint.
+	CgroupsPath string `json:"cgroupsPath"`
```
Comment (mrunalp) [inline runtime_config_linux.go]: @vishh It will be either CgroupsPath or Resources, right? We should either clarify that in the spec or have a layer of indirection.

**Comment by mrunalp** [comment]
@vishh I think that this should work with systemd as well. The only difference is the terminology AFAICS.

For e.g.
```
10:blkio:/system.slice/sshd.service
9:hugetlb:/
8:perf_event:/
7:devices:/system.slice/sshd.service
6:freezer:/
5:memory:/system.slice/sshd.service
4:net_cls,net_prio:/
3:cpuset:/
2:cpu,cpuacct:/system.slice/sshd.service
1:name=systemd:/system.slice/sshd.service

# mount
...
cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/usr/lib/systemd/systemd-cgroups-agent,name=systemd)
...
```

File: runtime_config_linux.go (hunk 3 above)
Comment (dqminh) [inline runtime_config_linux.go]: i agree that it should either be path or resources as it doesnt really make sense ( or even wrong ) to join with existing paths and then modify those resources.

Comment (vishh) [inline runtime_config_linux.go]: @mrunalp: How about creating the cgroups for the user and applying the limits expressed via `Resources`? 
@dqminh: Why would it be wrong? AFAIK the Spec doesn't try to limit how the existing Linux functionalities can be consumed right?

Comment (mrunalp) [inline runtime_config_linux.go]: @vishh There could be two interpretations. One is simple joining the cgroups (that were configured by some other tool or by another container) and other is using that path and then going in and modifying the values under that path. The first is necessary for us to support exec like functionality. Does that work for your use cases?

Comment (dqminh) [inline runtime_config_linux.go]: Hmm actually i withdraw my comment about enforcing the limitation of cgroupPaths vs resources. I think there can be usecases where having that capabilities is helpful ( like launching an agent container that modifies the cgroup resources at boot time and/or during the agent runtime maybe ? ).

I think that clarifying the order of resources modification is good enough i.e, if cgroups path is specified then we join the paths, otherwise we create new paths, and then apply the resources limit.

Comment (vishh) [inline runtime_config_linux.go]: @mrunalp: Your reasoning is sound! I see use cases for both `exec` and `resource updates`. Would making `Resources` optional resolve the issue?

Comment (wking) [inline runtime_config_linux.go]: On Tue, Sep 08, 2015 at 03:23:20PM -0700, Daniel, Dao Quang Minh wrote:
> I think that clarifying the order of resources modification is good
> enough i.e, if cgroups path is specified then we join the paths,
> otherwise we create new paths, and then apply the resources limit.

See also comments in #155 and #158, where we're having the same
discussion about namespaces.

Comment (mrunalp) [inline runtime_config_linux.go]: > @mrunalp: Your reasoning is sound! I see use cases for both exec and resource updates. Would making Resources optional resolve the issue? 

@vishh Sounds good to me!

**Comment by philips** [comment]
nit: reformat to one sentence per line please

**Comment by mrunalp** [comment]
Needs rebase as well.

**Comment by vishh** [comment]
@mrunalp @crosbymichael: WDYT of making cgroups path required? Without that, the options are cgroups can created under under root `/`, or under a user specified hierarchy (which requires configuration).

**Comment by mrunalp** [comment]
@vishh I think we could default to ` /` if cgroups path is not specified and a user specified path otherwise.

**Comment by vishh** [comment]
What will be the identifier we use for the cgroups? Will it be the base
name of the bundle?

**Comment by vbatts** [comment]
That is a common thought.

**Comment by mrunalp** [comment]
@vishh runc uses the --id or the base name if not specified.

**Comment by wking** [comment]
On Wed, Sep 09, 2015 at 04:55:45PM -0700, Vish Kannan wrote:
> What will be the identifier we use for the cgroups? Will it be the
> base name of the bundle?

Does this need to be spec'ed out?  There's not currently much language
around container IDs.  We just list them in state.json, and I've
pointed out that I'm not clear on why [1].  It makes sense that we'd
want some sort of handle for tracking “the same container” across a
checkpoint/restore migration (which is likely to change the PID), but
I don't see a need to bind the container ID in more tightly than that.
I think we should leave the cgroup/namespace name generation to the
runtime, and have consumers query (via the PID) to figure out the
cgroup/namespace name if they need it for something.

[1]: https://groups.google.com/a/opencontainers.org/d/msg/dev/z25xQsF3pHA/w0XqQhRyFwAJ
     Message-ID: <20150905041822.GA25638@odin.tremily.us>

**Comment by vishh** [comment]
@wking: My understanding is that it is recommended for runtimes to manage cgroups. But for a good dev experience, we can default to creating cgroups under `/` with an easy to use identifier (base name of the bundle).

**Comment by vishh** [comment]
@mrunalp @vbatts @dqminh Updated docs based on comments and rebased. PTAL.

**Comment by vbatts** [comment]
Though base of the bundle would not be unique enough for more than 1
container runtime. Let's leave this id for the runtime. Won't the paths be
referencable through the /run/opencontainers/ path anyways.

**Comment by vishh** [comment]
Yeah. The cgroup paths can be retrieved from the state.

**Comment by wking** [comment]
On Wed, Sep 09, 2015 at 05:13:32PM -0700, Vish Kannan wrote:
> My understanding is that it is recommended for runtimes to manage
> cgroups. But for a good dev experience, we can default to creating
> cgroups under `/` with an easy to use identifier (base name of the
> bundle).

If folks can spawn new containers without setting ‘cgroupsPath’, then
folks using a runtime shouldn't be guessing “/{container.id}” (or
whatever) as a cgroup path, they'll have to look it up for each
container (which I think is a good thing).

So since users will have to lookup a process's cgroups path anyway,
why should we bother specifying a rule to auto-generate paths?  If
different runtimes pick different rules for the default cgroup path,
what use-cases will suffer?

Wed, Sep 09, 2015 at 06:43:27PM -0700, Vish Kannan:
> Yeah. The cgroup paths can be retrieved from the state. 

Not directly.  Folks need to use the PID to get them (and that seems
fine to me), at least with the spec that landed via #87, see [1].

[1]: https://github.com/opencontainers/specs/pull/87#issuecomment-126123937

File: runtime-config-linux.md [hunk 4]
```diff
@@ -131,6 +131,58 @@ Also known as cgroups, they are used to restrict resource usage for a container
 cgroups provide controls to restrict cpu, memory, IO, pids and network for the container.
 For more information, see the [kernel cgroups documentation](https://www.kernel.org/doc/Documentation/cgroups/cgroups.txt).
 
+The path to the cgroups can to be specified in the Spec via `cgroupsPath`.
+`cgroupsPath` is expected to be relative to the cgroups mount point.
+If not specified, cgroups will be created under '/' with the base name of the bundle.
```
Comment (dqminh) [inline runtime-config-linux.md]: I think it's more reasonable / simpler to error out here rather than continuing and generate a path based on base name of the bundle. `cgroupsPath` should be a required data right ?

Comment (wking) [inline runtime-config-linux.md]: On Thu, Sep 10, 2015 at 01:54:07AM -0700, Daniel, Dao Quang Minh wrote:
> > +If not specified, cgroups will be created under '/' with the base name of the bundle.
> 
> I think it's more reasonable / simpler to error out here rather than
> continuing and generate a path based on base name of the
> bundle. `cgroupsPath` should be a required data right ?

I often want cgroups for my containers, but don't care what the path
is.  I think we do want the ability to tell the runtime “I don't care,
just make something up”, which is what leaving cgroupsPath out does.
But I'd rather leave the “make something up” algorithm up to the
runtime [1].  See also the “several containers from one bundle” use
case [2].

[1]: https://github.com/opencontainers/specs/pull/137#issuecomment-139103731
[2]: https://github.com/opencontainers/specs/pull/137#issuecomment-139078372

Comment (vishh) [inline runtime-config-linux.md]: I'm in favor of defaulting to sub-containers of root. Ping @crosbymichael @mrunalp @vbatts

Comment (wking) [inline runtime-config-linux.md]: On Thu, Sep 10, 2015 at 10:46:59AM -0700, Vish Kannan wrote:
> > +If not specified, cgroups will be created under '/' with the base name of the bundle.
> 
> I'm in favor of defaulting to sub-containers of root.

If anyone disagrees with *that*, I think it's a whole nother can of
worms.  I think the disagreement so far has been about what the
sub-container is going to be called, not where it will be rooted.

**Comment by mrunalp** [comment]
LGTM

**Comment by crosbymichael** [comment]
LGTM

File: runtime-config-linux.md [hunk 5]
```diff
@@ -131,6 +131,60 @@ Also known as cgroups, they are used to restrict resource usage for a container
 cgroups provide controls to restrict cpu, memory, IO, pids and network for the container.
 For more information, see the [kernel cgroups documentation](https://www.kernel.org/doc/Documentation/cgroups/cgroups.txt).
 
+The path to the cgroups can to be specified in the Spec via `cgroupsPath`.
+`cgroupsPath` is expected to be relative to the cgroups mount point.
+If not specified, cgroups will be created under '/'.
+Implementations of the Spec can choose to name cgroups in any manner.
```
Comment (wking) [inline runtime-config-linux.md]: This is only true if `cgroupsPath` is unset.  So maybe rephrase as “In that case, implementations can choose to name cgroups in any manner.” to tie it into the previous ”If not specified…” sentence.

Comment (vbatts) [inline runtime-config-linux.md]: I feel it's clear enough for now.

**Comment by vbatts** [comment]
LGTM
